xarray/core/dataset.py

```
@@ -4149,6 +4163,108 @@ def reset_index(
 
         return self._replace(variables, coord_names=coord_names, indexes=indexes)
 
+    def set_xindex(
+        self,
+        coord_names: Hashable | Sequence[Hashable],
+        index_cls: type[Index] | None = None,
+        **options,
+    ) -> Dataset:
+        """Set a new, Xarray-compatible index from one or more existing
+        coordinate(s).
+
+        Parameters
+        ----------
+        coord_names : str or list
+            Name(s) of the coordinate(s) used to build the index.
+            If several names are given, their order matters.
+        index_cls : subclass of :class:`~xarray.Index`, optional
+            The type of index to create. By default, try setting
+            a pandas (multi-)index from the supplied coordinates.
+        **options
+            Options passed to the index constructor.
+
+        Returns
+        -------
+        obj : Dataset
+            Another dataset, with this dataset's data and with a new index.
+
+        """
+        # the Sequence check is required for mypy
+        if is_scalar(coord_names) or not isinstance(coord_names, Sequence):
+            coord_names = [coord_names]
+
+        if index_cls is None:
+            if len(coord_names) == 1:
+                index_cls = PandasIndex
+            else:
+                index_cls = PandasMultiIndex
+        else:
+            if not issubclass(index_cls, Index):
+                raise TypeError(f"{index_cls} is not a subclass of xarray.Index")
+
+        invalid_coords = set(coord_names) - self._coord_names
+
+        if invalid_coords:
+            raise ValueError(f"those coordinates don't exist: {invalid_coords}")
```

Let's raise a nicer error if the variables are in `data_vars`